### PR TITLE
Trigger should always perform message batching

### PR DESCRIPTION
**Pull request by carlpulley-da**
Previously, triggers would only batch messages when rule evaluation started to back pressure. This PR introduces an optimisation whereby we always batch trigger messages.

Trigger context logging refactored to better manage message batching based contexts.

### CODE REVIEW

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 1]
```diff
@@ -799,32 +799,18 @@ private[lf] class Runner private (
...
-            }
-          },
-        ) { case (ctx @ Ctx(batchContext, stack, _), Ctx(context, value, _)) =>
-          batchContext
-            .groupWith(context)
-            .logDebug("Batching message for rule evaluation", "index" -> (stack.length + 1))
-
-          ctx.copy(value = stack :+ value)
+        .groupedWithin(triggerConfig.maximumBatchSize.toInt, triggerConfig.batchingDuration)
+        .collect { case batch @ ctx +: _ =>
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:810]: We intentionally drop empty lists of message batches (in theory, they should never be emitted!).

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala [hunk 2]
```diff
@@ -799,32 +799,24 @@ private[lf] class Runner private (
...
+        .groupedWithin(triggerConfig.maximumBatchSize.toInt, triggerConfig.batchingDuration)
+        // We drop empty message batches only here
+        .collect { case batch @ ctx +: remaining =>
+          ctx.context.childSpan("batch", "size" -> batch.size) {
+            implicit triggerContext: TriggerLogContext =>
+              ctx.copy(
+                context = triggerContext.groupWith(remaining.map(_.context): _*),
+                value = SList(FrontStack(batch.map(_.value): _*)),
+              )
+          }
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/Runner.scala:818]: The child `batch` logging context is defined to have all batched message contexts as parents.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerLogContext.scala [hunk 3]
```diff
@@ -66,8 +71,15 @@ final class TriggerLogContext private (
     f(new TriggerLogContext(loggingContext, entries ++ additionalEntries, span.childSpan(name)))
   }
 
-  def groupWith(context: TriggerLogContext): TriggerLogContext = {
-    new TriggerLogContext(loggingContext, entries, span.groupWith(context.span))
+  def groupWith(contexts: TriggerLogContext*): TriggerLogContext = {
+    val groupEntries = contexts.foldLeft(entries.toSet) { case (entries, context) =>
+      entries ++ context.entries.toSet
+    }
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerLogContext.scala:77]: We work with sets here to reduce risk of duplicate entries in the logs

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerLogContext.scala [hunk 4]
```diff
@@ -126,7 +138,7 @@ object TriggerLogContext {
   private[trigger] final case class TriggerLogSpan(
       path: BackStack[String],
       id: UUID = UUID.randomUUID(),
-      parent: Option[UUID] = None,
+      parent: Set[UUID] = Set.empty,
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerLogContext.scala:141]: For message batching, we allow children to have multiple parents.

File: triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRunnerConfig.scala [hunk 5]
```diff
@@ -48,6 +49,7 @@ object TriggerRunnerConfig {
       // 256 here comes from the default ExecutionContext.
       submissionFailureQueueSize = 256 + parallelism,
       maximumBatchSize = 1000,
+      batchingDuration = 250.milliseconds,
```
Comment (carlpulley-da) [inline triggers/runner/src/main/scala/com/digitalasset/daml/lf/engine/trigger/TriggerRunnerConfig.scala:52]: We batch messages into groups of 1000 every 250ms. Whichever limit is hit first determines when trigger evaluation occurs for the batch.

**Review by remyhaemmerle-da — APPROVED** [review]
Nice !

**Pull request merged**